### PR TITLE
samples to show how to switch to another region when service unavailable

### PR DESCRIPTION
**Pull request by yulin-li**
## Purpose
samples to show how to switch to another region when service unavailable for TTS, including standard voice and custom voice

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

### CODE REVIEW

File: samples/csharp/sharedcontent/console/speech_synthesis_samples.cs [hunk 1]
```diff
@@ -850,5 +850,72 @@ public static async Task SynthesisGetAvailableVoicesAsync()
...
+                    || details.ErrorCode == CancellationErrorCode.ServiceTimeout 
+                    || details.ErrorDetails.Contains("Error code: 1007"))
+                {
+                    // Synthesize using platform voice. The SecondarySubscription could be same with primary one
+                    synthesisResult = await SynthesizeOnceAsyncInternal("SecondarySubscriptionRegion", "SecondarySubscriptionRegion");
+                }
+            }
+        }
     }
 }
```
Comment (ForrestGumb) [inline samples/csharp/sharedcontent/console/speech_synthesis_samples.cs:921]: Please add a separate sample for Custom Voice fallback to Custom Voice.
        // Custom voice with backup.
        // If custom voice service is unavailable, you can fallback to custom voice service in another region.
        public static async Task SynthesizeOnceUseCustomVoiceToSpeakerAsyncSwitchSecondaryRegion()
        {
            // Create a speech resource with primary subscription key and service region.
            // Also create a speech resource with secondary subscription key and service region.
            // Copy Custom Voice model from primary region to secondary region and deploy.
            SpeechSynthesisResult synthesisResult = await SynthesizeOnceAsyncInternal("PrimarySubscriptionKey",
                "PrimarySubscriptionRegion", "YourEndpointIdOnPrimaryRegion", "YourVoiceName");
            if (synthesisResult.Reason == ResultReason.Canceled)
            {
                SpeechSynthesisCancellationDetails details = SpeechSynthesisCancellationDetails.FromResult(synthesisResult);
                if (details.ErrorCode == CancellationErrorCode.ConnectionFailure
                    || details.ErrorCode == CancellationErrorCode.ServiceUnavailable
                    || details.ErrorCode == CancellationErrorCode.ServiceTimeout
                    || details.ErrorDetails.Contains("Error code: 1007"))
                {
                    // Synthesize using platform voice. The SecondarySubscription could be same with primary one
                    synthesisResult = await SynthesizeOnceAsyncInternal("SecondarySubscriptionRegion", "SecondarySubscriptionRegion",
                        "YourEndpointIdOnSecondaryRegion", "YourVoiceName");
                }
            }
        }

File: samples/csharp/sharedcontent/console/speech_synthesis_samples.cs [hunk 2]
```diff
@@ -850,5 +850,72 @@ public static async Task SynthesisGetAvailableVoicesAsync()
...
+                config.EndpointId = endpointId;
+            }
+
+            if (!string.IsNullOrEmpty(voiceName))
+            {
+                config.SpeechSynthesisVoiceName = voiceName;
+            }
+
+            // Creates a speech synthesizer using the default speaker as audio output.
+            using (var synthesizer = new SpeechSynthesizer(config, null as AudioConfig))
```
[the remark this answers is not in the document]
Comment (yulin-li) [reply samples/csharp/sharedcontent/console/speech_synthesis_samples.cs:870]: Removed, thx.

**Review by ForrestGumb — APPROVED** [review]

File: samples/csharp/sharedcontent/console/speech_synthesis_samples.cs [hunk 3]
```diff
@@ -850,5 +850,93 @@ public static async Task SynthesisGetAvailableVoicesAsync()
...
+        // Speech synthesis with backup subscription region.
+        public static async Task SynthesizeOnceToSpeakerAsyncSwitchSecondaryRegion()
+        {
+            // Create a speech resource with primary subscription key and service region.
+            // Also create a speech resource with secondary subscription key and service region
+            SpeechSynthesisResult synthesisResult = await SynthesizeOnceAsyncInternal("PrimarySubscriptionKey", "PrimarySubscriptionRegion" );
+            if (synthesisResult.Reason == ResultReason.Canceled)
+            {
+                SpeechSynthesisCancellationDetails details = SpeechSynthesisCancellationDetails.FromResult(synthesisResult);
+                if (details.ErrorCode == CancellationErrorCode.ConnectionFailure
```
Comment (BrianMouncer) [inline samples/csharp/sharedcontent/console/speech_synthesis_samples.cs:888]: >ConnectionFailure [](http://example.com/codeflow?start=63&length=17)

The other two seem obvious, but is ConnectionFailure a little too generic to assume you should switch regions?

**Review by BrianMouncer — APPROVED** [review]
:shipit:

File: samples/csharp/sharedcontent/console/speech_synthesis_samples.cs (hunk 1 above)
Comment (yulin-li) [reply samples/csharp/sharedcontent/console/speech_synthesis_samples.cs:942]: added, thx

File: samples/csharp/sharedcontent/console/speech_synthesis_samples.cs (hunk 3 above)
Comment (yulin-li) [reply samples/csharp/sharedcontent/console/speech_synthesis_samples.cs:888]: Yes, connection failure is too generic. Some client issue causes connection failure. I think the sample is used in such scenario: the primary region worked yesterday, and don't work today without any code change. Then the failure reason is mostly a service issue

**Pull request merged**